// language/tlang/regallocate.ts

import * as m from './mipscode';
import * as util from './util';
import * as ic from './intermediatecode';
import * as t from './tac';

//NOTE, THIS FUNCTION WILL MODIFY INPUT (codelines)
export function regalloc(codelines: Array<ic.CodeLine>, regvinfer: Array<ic.CodeLineRegInfoInferences>, maxtmpregnum: number): { regmap: Map<number, number>, tmpinstack: Map<number, number> } {
    let rig = new RIG(), clen = codelines.length;
    for (let i = 0; i < clen; ++i) {
        rig.addRelateRegs([...regvinfer[i].bottom_live]);
        rig.addRelateRegs([...regvinfer[i].top_live]);
    }
    let allo = rig.allocate(), fpoffset = 0, stackoffsetmap = new Map<number, number>(), tmpregnum = maxtmpregnum + 1;
    while (!allo.succeed) {
        let snum = allo.spill, i = 0;
        //new assigned reg for spilled reg cannot be spilled again (no good solution to this conflict yet)
        if (snum > maxtmpregnum) throw new Error('undefined behavior, spilled reigster cannot be spilled again: ' + snum);
        let mloc = -4 * fpoffset++;
        stackoffsetmap.set(snum, mloc);
        rig.removeReg(snum);
        for (; i < codelines.length; ++i) {
            //remove spilled tmp-register in every lines
            let reginfer = regvinfer[i], cl = codelines[i];
            reginfer.bottom_live.delete(snum);
            reginfer.top_live.delete(snum);
            let tac = cl.tac;
            if (tac.readReg(snum)) {
                let newreg = tmpregnum++;
                let newcl = new ic.CodeLine(new t.TAC_lw(util.TMP_REGS_FP, mloc, newreg));
                let newreginfer = new ic.CodeLineRegInfoInferences(0);
                newreginfer.top_live = new Set<number>(reginfer.top_live);
                newreginfer.bottom_live = new Set<number>(reginfer.top_live).add(newreg);
                codelines.splice(i, 0, newcl);
                regvinfer.splice(i++, 0, newreginfer);
                if (cl.label != null) {
                    let lb = cl.label;
                    cl.label = null;
                    newcl.label = lb;
                    lb.owner = newcl;
                }
                tac.replReadReg(snum, newreg);
                rig.addRelateRegsForOne(newreg, [...reginfer.top_live]);
                reginfer.top_live.add(newreg);
            }
            if (tac.writeReg(snum)) {
                let newreg = tmpregnum++;
                let newcl = new ic.CodeLine(new t.TAC_sw(util.TMP_REGS_FP, mloc, newreg));
                let newreginfer = new ic.CodeLineRegInfoInferences(0);
                newreginfer.bottom_live = new Set<number>(reginfer.bottom_live);
                newreginfer.top_live = new Set<number>(reginfer.bottom_live).add(newreg);
                codelines.splice(++i, 0, newcl);
                regvinfer.splice(i, 0, newreginfer);
                //ATTENTION, IF WE HAVE INSTRUCTION TO WRITE INTO SOME REGISTER AND BRANCH TO OTHER PC SIMUTANEOUSLY, WE NEED PROCESS THE 'BRANCH TARGET' HERE
                //MIPS DOES NOT HAVE THESE INSTRUCTIONS
                tac.replWriteReg(snum, newreg);
                rig.addRelateRegsForOne(newreg, [...reginfer.bottom_live]);
                reginfer.bottom_live.add(newreg);
            }
        }
        allo = rig.allocate(); //try allocate register again
    }
    return { regmap: allo.map, tmpinstack: stackoffsetmap };
}

//register interference graph
class RIG {
    private _map: Map<number, Set<number>>;

    removeReg(regnum: number): this {
        let s = this._map.get(regnum);
        if (s != null) {
            this._map.delete(regnum);
            for (let r of s)
                this._map.get(r).delete(regnum);
        }
        return this;
    }
    //complete linkage
    addRelateRegs(regnums: Array<number>): this {
        let rlen = regnums.length;
        if (rlen === 1) {
            this._getOrCreateSet(regnums[0]);
        }
        else {
            for (let i = 0; i < rlen - 1; ++i) {
                for (let j = i + 1; j < rlen; ++j) {
                    this._addRelatedRegs(regnums[i], regnums[j]);
                }
            }
        }
        return this;
    }
    //1->many link
    addRelateRegsForOne(regnum: number, regnums: Array<number>): this {
        if (regnums.length === 0) this._getOrCreateSet(regnum);
        else {
            let rlen = regnums.length;
            for (let i = 0; i < rlen; ++i) {
                this._addRelatedRegs(regnums[i], regnum);
            }
        }
        return this;
    }
    private _addRelatedRegs(regnum1: number, regnum2: number): this {
        if (regnum1 !== regnum2) {
            this._getOrCreateSet(regnum1).add(regnum2);
            this._getOrCreateSet(regnum2).add(regnum1);
        }
        return this;
    }
    private _getOrCreateSet(regnum: number) {
        let s = this._map.get(regnum);
        if (s == null) {
            s = new Set<number>();
            this._map.set(regnum, s);
        }
        return s;
    }
    allocate(): { succeed: boolean, spill?: number, map?: Map<number, number> } {
        let removables = new Array<number>(), unremovable = new Set<number>(), rarr = new Array<number>(), tmpcount = this._map.size, tmpmap = new Map<number, Set<number>>();
        for (let x of this._map) {
            if (x[1].size < regCount) removables.push(x[0]);
            else unremovable.add(x[0]);
            let tset = new Set<number>();
            //clone the set
            for (let n of x[1]) tset.add(n);
            tmpmap.set(x[0], tset);
        }
        while (true) {
            while (removables.length > 0) {
                let rnum = removables.pop();
                rarr.push(rnum);
                for (let n of tmpmap.get(rnum)) {
                    let s = tmpmap.get(n);
                    s.delete(rnum);
                    if (s.size < regCount && unremovable.has(n)) {
                        removables.push(n);
                        unremovable.delete(n);
                    }
                }
                tmpmap.delete(rnum);
            }
            if (rarr.length === tmpcount)
                break;
            else {
                let maxsize = -1, maxnum = -1;
                for (let t of tmpmap) {
                    if (maxsize < t[1].size) {
                        maxsize = t[1].size;
                        maxnum = t[0];
                    }
                }
                if (maxnum === -1 || maxsize === -1) throw new Error('defensive code, impossible code path');
                removables.push(maxnum);
                unremovable.delete(maxnum);
            }
        }
        if (tmpmap.size !== 0) throw new Error('defensive code, impossible code path ' + tmpmap.size);
        let regmap = new Map<number, number>(), i = rarr.length - 1;
        while (i >= 0) {
            let rnum = rarr[i];
            let nei = this._map.get(rnum);
            let assignedneibor = new Array<number>(), regs = new Array<boolean>(regCount);
            for (let j = i + 1; j < rarr.length; ++j) {
                if (nei.has(rarr[j]))
                    assignedneibor.push(rarr[j]);
            }
            for (let n of assignedneibor) regs[regmap.get(n)] = true;
            let j = 0;
            for (; j < regCount; ++j) {
                if (regs[j] !== true) {
                    //assign the register to temporary
                    regmap.set(rnum, j);
                    break;
                }
            }
            //if no valid assignment
            if (j === regCount)
                return { succeed: false, spill: rnum };
            --i;
        }
        return { succeed: true, map: regmap };
    }
    constructor() {
        this._map = new Map<number, Set<number>>();
    }
}

export function toMIPSReg(tmpnum: number, regmap: Map<number, number>): m.MIPSRegister {
    if (tmpnum === util.TMP_REGS_FP) return m.REGS.fp;
    else if (!regmap.has(tmpnum)) throw new Error('undefined regnum for given temporary: ' + tmpnum);
    else return regnumToMIPSReg(regmap.get(tmpnum));
};

const regCount = 3;

export function regnumToMIPSReg(regnum: number): m.MIPSRegister {
    switch (regnum) {
        case 0: return m.REGS.t0;
        case 1: return m.REGS.t1;
        case 2: return m.REGS.t2;
        case 3: return m.REGS.t3;
        default: throw new Error('invalid regnum: ' + regnum);
    }
}
