// productions/index.js
"use strict";
const prod = require('./production');
const _ = require('lodash');
let chA = 'A'.charCodeAt(0);
let chZ = 'Z'.charCodeAt(0);
function strAsNonTerminal(str) {
    let ch = str.charCodeAt(0);
    return chA <= ch && ch <= chZ;
}
;
function createProduction(lstr, rstr) {
    lstr = lstr.trim();
    rstr = rstr.trim();
    if (lstr.length === 0)
        throw new Error('left-hand-side is empty');
    if (lstr.indexOf(' ') >= 0)
        throw new Error('left-hand-side contains more than one symbol: ' + lstr);
    if (!strAsNonTerminal(lstr))
        throw new Error('left-hand-side is not a non-terminal symbol: ' + lstr);
    if (_.startsWith(lstr, prod.ProdSet.preservedNonTerminalPrefix))
        throw new Error('non-terminal cannot start with preserved prefix: ' + lstr);
    let lsymbol = new prod.Symbol(false, lstr);
    let rsymbols = rstr.split(' ').filter(str => str.length > 0).map(str => {
        if (_.startsWith(str, prod.ProdSet.preservedNonTerminalPrefix))
            throw new Error('non-terminal cannot start with preserved prefix: ' + str);
        return new prod.Symbol(!strAsNonTerminal(str), str);
    });
    return new prod.Production(lsymbol, rsymbols);
}
function createProduction2(prodstr) {
    let idxarrow = prodstr.indexOf('->');
    if (idxarrow < 0)
        throw new Error('production must has a -> :' + prodstr);
    let lstr = prodstr.slice(0, idxarrow), rstr = prodstr.slice(idxarrow + 2);
    return createProduction(lstr, rstr);
}
function createMultipleProductions(mprodstr) {
    let idxarrow = mprodstr.indexOf('->');
    if (idxarrow < 0)
        throw new Error('production must has a -> :' + mprodstr);
    let lstr = mprodstr.slice(0, idxarrow).trim();
    return mprodstr.slice(idxarrow + 2).trim().split('|').map(rstr => createProduction(lstr, rstr));
}
function createProdSet(mprodarr) {
    let prods = [];
    for (let mprod of mprodarr) {
        for (var p of createMultipleProductions(mprod))
            prods.push(p);
    }
    return new prod.ProdSet(prods);
}
exports.createProdSet = createProdSet;
var production_1 = require('./production');
exports.Production = production_1.Production;
exports.ProdSet = production_1.ProdSet;
exports.Symbol = production_1.Symbol;
